File: frontend/frontend/lib/models/hogar.dart
// frontend/lib/models/hogar.dart

class Hogar {
  final int idHogar;
  final String nombre;
  final String icono;
  final int totalMiembros;
  final String rol; // 'admin', 'miembro', 'invitado'
  
  Hogar({
    required this.idHogar,
    required this.nombre,
    required this.icono,
    required this.totalMiembros,
    required this.rol,
  });
  
  factory Hogar.fromJson(Map<String, dynamic> json) {
    return Hogar(
      idHogar: json['id_hogar'],
      nombre: json['nombre'],
      icono: json['icono'] ?? 'home',
      totalMiembros: json['total_miembros'] ?? 0,
      rol: json['rol'] ?? 'miembro',
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      'id_hogar': idHogar,
      'nombre': nombre,
      'icono': icono,
      'total_miembros': totalMiembros,
      'rol': rol,
    };
  }
}

class HogarDetalle {
  final int idHogar;
  final String nombre;
  final String icono;
  final String codigoInvitacion;
  final List<Miembro> miembros;
  
  HogarDetalle({
    required this.idHogar,
    required this.nombre,
    required this.icono,
    required this.codigoInvitacion,
    required this.miembros,
  });
  
  factory HogarDetalle.fromJson(Map<String, dynamic> json) {
    return HogarDetalle(
      idHogar: json['id_hogar'],
      nombre: json['nombre'],
      icono: json['icono'] ?? 'home',
      codigoInvitacion: json['codigo_invitacion'],
      miembros: (json['miembros'] as List)
          .map((m) => Miembro.fromJson(m))
          .toList(),
    );
  }
}

class Miembro {
  final String userId;
  final String apodo;
  final String rol;
  final String fechaUnion;
  
  Miembro({
    required this.userId,
    required this.apodo,
    required this.rol,
    required this.fechaUnion,
  });
  
  factory Miembro.fromJson(Map<String, dynamic> json) {
    return Miembro(
      userId: json['user_id'],
      apodo: json['apodo'] ?? 'Usuario',
      rol: json['rol'],
      fechaUnion: json['fecha_union'],
    );
  }
}

File: backend/schemas/hogar.py
# backend/schemas/hogar.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional


class HogarBase(BaseModel):
    """Base schema for household."""
    nombre: str = Field(..., min_length=1, max_length=200, description="Household name")
    icono: str = Field(default='home', max_length=50, description="Icon identifier")


class HogarCreate(HogarBase):
    """Schema for creating a new household."""
    pass


class HogarUpdate(BaseModel):
    """Schema for updating a household."""
    nombre: Optional[str] = Field(None, min_length=1, max_length=200)
    icono: Optional[str] = Field(None, max_length=50)


class MiembroInfo(BaseModel):
    """Schema for household member information."""
    user_id: str
    rol: str
    apodo: Optional[str] = None
    fecha_union: datetime
    
    class Config:
        from_attributes = True


class HogarSchema(HogarBase):
    """Complete household schema with all fields."""
    id_hogar: int
    created_by: str
    fecha_creacion: datetime
    codigo_invitacion: str
    miembros_count: int = Field(default=0, description="Number of members")
    mi_rol: Optional[str] = Field(None, description="Current user's role in this household")
    
    class Config:
        from_attributes = True


class HogarDetalle(HogarSchema):
    """Detailed household schema including members list."""
    miembros: list[MiembroInfo] = []
    
    class Config:
        from_attributes = True


class HogarMiembroBase(BaseModel):
    """Base schema for household membership."""
    fk_hogar: int
    user_id: str
    rol: str = Field(default='miembro', pattern='^(admin|miembro|invitado)$')
    apodo: Optional[str] = Field(None, max_length=100)


class HogarMiembroCreate(BaseModel):
    """Schema for adding a member to household (usually via invitation)."""
    codigo_invitacion: str = Field(..., min_length=8, max_length=8)
    apodo: Optional[str] = Field(None, max_length=100)


class HogarMiembroUpdate(BaseModel):
    """Schema for updating member information."""
    rol: Optional[str] = Field(None, pattern='^(admin|miembro|invitado)$')
    apodo: Optional[str] = Field(None, max_length=100)


class HogarMiembroSchema(HogarMiembroBase):
    """Complete member schema."""
    id_miembro: int
    fecha_union: datetime
    
    class Config:
        from_attributes = True


class InvitacionResponse(BaseModel):
    """Response when generating invitation code."""
    codigo_invitacion: str
    hogar_nombre: str
    expira_en: Optional[str] = Field(None, description="Optional expiration info")
